database/mssql/scripts/dbscripts/PSP_PIMS_S24_00/Alter Down/01_PSP_PIMS_S24_00_Alter_Down.sql

```diff
@@ -0,0 +1,3335 @@
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop unique constraint dbo.PIMS_PRSCRC_H_UK
+PRINT N'Drop unique constraint dbo.PIMS_PRSCRC_H_UK'
+GO
+ALTER TABLE [dbo].[PIMS_PROPERTY_RESEARCH_FILE_HIST]
+	DROP CONSTRAINT [PIMS_PRSCRC_H_UK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.PIMS_PRSCRC_H_PK
+PRINT N'Drop primary key constraint dbo.PIMS_PRSCRC_H_PK'
+GO
+ALTER TABLE [dbo].[PIMS_PROPERTY_RESEARCH_FILE_HIST]
+	DROP CONSTRAINT [PIMS_PRSCRC_H_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.PRSCRC_PK
+PRINT N'Drop primary key constraint dbo.PRSCRC_PK'
+GO
+ALTER TABLE [dbo].[PIMS_PROPERTY_RESEARCH_FILE]
+	DROP CONSTRAINT [PRSCRC_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop table dbo.PIMS_PROPERTY_RESEARCH_FILE
+PRINT N'Drop table dbo.PIMS_PROPERTY_RESEARCH_FILE'
+GO
+DROP TABLE [dbo].[PIMS_PROPERTY_RESEARCH_FILE]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop unique constraint dbo.PIMS_RESRCH_H_UK
+PRINT N'Drop unique constraint dbo.PIMS_RESRCH_H_UK'
+GO
+ALTER TABLE [dbo].[PIMS_RESEARCH_FILE_HIST]
+	DROP CONSTRAINT [PIMS_RESRCH_H_UK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.PIMS_RESRCH_H_PK
+PRINT N'Drop primary key constraint dbo.PIMS_RESRCH_H_PK'
+GO
+ALTER TABLE [dbo].[PIMS_RESEARCH_FILE_HIST]
+	DROP CONSTRAINT [PIMS_RESRCH_H_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.RESRCH_PK
+PRINT N'Drop primary key constraint dbo.RESRCH_PK'
+GO
+ALTER TABLE [dbo].[PIMS_RESEARCH_FILE]
+	DROP CONSTRAINT [RESRCH_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.RSRCHS_PK
+PRINT N'Drop primary key constraint dbo.RSRCHS_PK'
+GO
+ALTER TABLE [dbo].[PIMS_RESEARCH_FILE_STATUS_TYPE]
+	DROP CONSTRAINT [RSRCHS_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.BCAJUR_PK
+PRINT N'Drop primary key constraint dbo.BCAJUR_PK'
+GO
+ALTER TABLE [dbo].[BCA_JURISDICTION]
+	DROP CONSTRAINT [BCAJUR_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.BCASAR_PK
+PRINT N'Drop primary key constraint dbo.BCASAR_PK'
+GO
+ALTER TABLE [dbo].[BCA_ASSESSMENT_AREA]
+	DROP CONSTRAINT [BCASAR_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop primary key constraint dbo.VOLUTY_PK
+PRINT N'Drop primary key constraint dbo.VOLUTY_PK'
+GO
+ALTER TABLE [dbo].[PIMS_VOLUME_UNIT_TYPE]
+	DROP CONSTRAINT [VOLUTY_PK]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Drop sequence dbo.PIMS_RFILE_NUMBER_SEQ
+PRINT N'Drop sequence dbo.PIMS_RFILE_NUMBER_SEQ'
+GO
+DROP SEQUENCE [dbo].[PIMS_RFILE_NUMBER_SEQ]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Alter table dbo.PIMS_PROPERTY
+PRINT N'Alter table dbo.PIMS_PROPERTY'
+GO
+ALTER TABLE [dbo].[PIMS_PROPERTY] ALTER COLUMN [PROPERTY_AREA_UNIT_TYPE_CODE] nvarchar(20) NOT NULL
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+ALTER TABLE [dbo].[PIMS_PROPERTY]
+	ADD [PROPERTY_TENURE_TYPE_CODE] nvarchar(20) NOT NULL, 
```

Comment:
This causes the following in the alter down:

Alter table dbo.PIMS_PROPERTY
Msg 4901, Level 16, State 1, Server 6c858a0849fd, Line 1
ALTER TABLE only allows columns to be added that can contain nulls, or have a DEFAULT definition specified, or the column being added is an identity or timestamp column, or alternatively if none of the previous conditions are satisfied the table must be empty to allow addition of this column. Column 'PROPERTY_TENURE_TYPE_CODE' cannot be added to non-empty table 'PIMS_PROPERTY' because it does not satisfy these conditions.

I'm nervous that we are still getting these, does Aqua Data Studio handle these automatically or not? If it doesn't, we need to be more careful when changing fields to non-null on existing tables. The only reason I am seeing this is because I have data in this table. If this was on a table we weren't using yet, like research, we might not find these at all, which could lead to issues in a future deployment.

Reply:
ADS handles it automatically but it gives you some options as to what to do in these situations.  Usually it's a choice of "allow nulls", "ignore", "inseert custom value", or choose from a value in a related (by a portion of the name) column.  In this particular case, I probably chose "ignore".

Reply:
Ok, I'll keep an eye out for the fix to the alter down.

Reply:
Was this the only instance of missing defaults that you saw?